packages/flutter/lib/src/foundation/synchronous_future.dart

```diff
@@ -42,8 +42,17 @@ class SynchronousFuture<T> implements Future<T> {
   }
 
   @override
-  Future<T> timeout(Duration timeLimit, { Future<T> onTimeout() }) => new Completer<T>().future;
+  Future<T> timeout(Duration timeLimit, { dynamic onTimeout() }) => new Completer<T>().future;
 
   @override
-  Future<T> whenComplete(Future<T> action()) => action();
+  Future<T> whenComplete(dynamic action()) {
+    try {
+      dynamic result = action();
+      if (result is Future)
```

Comment:
I'm surprised the analyzer didn't complain about the missing type here, but I don't see value in being explicit about the `<dynamic>` so that's fine I guess!